util/gen_pet_names.ts

```diff
@@ -1,71 +1,115 @@
 import path from 'path';
 
-import fetch from 'node-fetch';
-
-import { CoinachWriter } from './coinach';
 import { getCnTable, getKoTable } from './csv_util';
+import { OutputFileAttributes, XivApi } from './xivapi';
 
-const _OUTPUT_FILE = 'pet_names.ts';
+const _PET_NAMES: OutputFileAttributes = {
+  outputFile: 'resources/pet_names.ts',
+  type: 'PetData',
+  header: `import { Lang } from './languages';
 
-const keys = ['Name'];
-const localeKeys = [...keys, 'Name_de', 'Name_fr', 'Name_ja'];
+type PetData = {
+  [name in Lang]: readonly string[];
+};
+`,
+  asConst: false,
+};
 
-type XivapiResult = {
-  Results: {
-    [key: number]: Record<'Name' | 'Name_de' | 'Name_fr' | 'Name_ja', string>;
-  };
+const _ENDPOINT = 'Pet';
+
+const _COLUMNS = [
+  'ID',
+  'Name_de',
+  'Name_en',
+  'Name_fr',
+  'Name_ja',
+];
+
+const _LOCALE_TABLE = 'Pet';
+
+const _LOCALE_COLUMNS = ['Name'];
+
+type ResultPet = {
+  Name_de: string | null;
+  Name_en: string | null;
+  Name_fr: string | null;
+  Name_ja: string | null;
 };
 
-const fetchXivapi = async () => {
-  const url = `https://xivapi.com/Pet?columns=${localeKeys.join(',')}`;
-  const response = await fetch(url);
-  const json = (await response.json()) as XivapiResult;
-  return json.Results;
+type XivApiPet = {
+  [key: number]: ResultPet;
 };
 
-const normalize = (
-  content: XivapiResult['Results'],
-): Record<'en' | 'de' | 'fr' | 'ja', string[]> => {
-  const result: Record<'en' | 'de' | 'fr' | 'ja', string[]> = {
-    // cactbot-ignore-missing-translations
-    en: [],
+type OutputPetNames = {
+  cn: string[];
+  de: string[];
+  en: string[];
+  fr: string[];
+  ja: string[];
+  ko: string[];
+};
+
+const fetchLocaleCsvTables = async () => {
+  const cnPet = await getCnTable(_LOCALE_TABLE, _LOCALE_COLUMNS);
+  const koPet = await getKoTable(_LOCALE_TABLE, _LOCALE_COLUMNS);
+  return {
+    cn: cnPet,
+    ko: koPet,
+  };
+};
+
+const assembleData = async (apiData: XivApiPet): Promise<OutputPetNames> => {
+  // This isn't really a locale object, and ordering is alpha in the current file, so:
+  // eslint-disable-next-line rulesdir/cactbot-locale-order
+  const formattedData: OutputPetNames = {
+    cn: [],
     de: [],
+    en: [],
     fr: [],
     ja: [],
+    ko: [],
   };
-  for (const [, data] of Object.entries(content)) {
-    // skip empty names and duplicates
-    if (data.Name === '' || result.en.includes(data.Name))
+
+  for (const pet of Object.values(apiData)) {
+    // If no en name (or if duplicate), skip processing
+    if (pet.Name_en === null || pet.Name_en === '' || formattedData.en.includes(pet.Name_en))
       continue;
+    formattedData.en.push(pet.Name_en);
 
-    result.en.push(data.Name);
-    result.de.push(data.Name_de);
-    result.fr.push(data.Name_fr);
-    result.ja.push(data.Name_ja);
+    // If there's an en name, the locale names will also be there.
+    // But Typescript doesn't know that.
```

Comment:
Could/should this instead be using a discriminated union such as this?

```ts
type ResultPet = {
  Name_en: null;
} | {
  Name_de: string;
  Name_en: string;
  Name_fr: string;
  Name_ja: string;
};
```

(Can't test right now to make sure this is correct/works, but it should?)

Reply:
> Could/should this instead be using a discriminated union such as this?

It should work, at least in terms of how the data is currently stored and returned to us.  But given that the data flows from Coinach/Lumina/? => xivapi's transformation scripts => api => fetch => stringify before we start working with it, I felt we should apply a low-trust typing philosophy and treat any data field other than the actual ID column of a table as potentially null to force explicit type checks on everything (same reason why every number field we import is a string | number).

Reply:
In that case, the comment here doesn't really make sense?